src/main/java/org/apache/sysml/api/mlcontext/Matrix.java

```diff
@@ -146,12 +207,38 @@ public MatrixObject toMatrixObject() {
 	}
 
 	/**
-	 * Obtain the matrix as a {@code BinaryBlockMatrix}
+	 * Obtain the matrix as a {@code JavaPairRDD<MatrixIndexes, MatrixBlock>}
 	 *
-	 * @return the matrix as a {@code BinaryBlockMatrix}
+	 * @return the matrix as a {@code JavaPairRDD<MatrixIndexes, MatrixBlock>}
 	 */
-	public BinaryBlockMatrix toBinaryBlockMatrix() {
-		return MLContextConversionUtil.matrixObjectToBinaryBlockMatrix(matrixObject, sparkExecutionContext);
+	public JavaPairRDD<MatrixIndexes, MatrixBlock> toBinaryBlocks() {
+		if (binaryBlocks != null) {
+			return binaryBlocks;
+		} else if (matrixObject != null) {
+			binaryBlocks = MLContextConversionUtil.matrixObjectToBinaryBlocks(matrixObject, sparkExecutionContext);
+			MatrixCharacteristics mc = matrixObject.getMatrixCharacteristics();
+			matrixMetadata = new MatrixMetadata(mc);
+			return binaryBlocks;
+		}
+		throw new MLContextException("No binary blocks or MatrixObject found");
+	}
+
+	/**
+	 * Obtain the matrix as a {@code MatrixBlock}
+	 *
+	 * @return the matrix as a {@code MatrixBlock}
+	 */
+	public MatrixBlock toMatrixBlock() {
+		if (matrixMetadata == null) {
+			throw new MLContextException("Matrix metadata required to convert binary blocks to a MatrixBlock.");
+		}
+		if (binaryBlocks != null) {
+			return MLContextConversionUtil.binaryBlocksToMatrixBlock(binaryBlocks, matrixMetadata);
+		} else if (matrixObject != null) {
+			toBinaryBlocks();
+			return MLContextConversionUtil.binaryBlocksToMatrixBlock(binaryBlocks, matrixMetadata);
```

Comment:
Could you use toBinaryBlocks() instead of binaryBlocks here? It would have the same effect but would make it more explicit without knowing that toBinaryBlocks also modifies the state.